tests/planner/ompl/test_OMPLSimplifier.cpp

```diff
@@ -0,0 +1,189 @@
+  aikido::constraint::TestablePtr boundsConstraint = aikido::constraint::createTestableBounds(stateSpace);
+  aikido::constraint::TestablePtr collConstraint = std::make_shared<MockTranslationalRobotConstraint>(
+        stateSpace, Eigen::Vector3d(-0.1, -0.1, -0.1),
+        Eigen::Vector3d(0.1, 0.1, 0.1));
+
+  auto simplifiedPair = aikido::planner::ompl::simplifyOMPL(
+      stateSpace, interpolator, std::move(dmetric), std::move(sampler), 
+      std::move(collConstraint), std::move(boundsConstraint), std::move(boundsProjection),
+      0.1, 5.0, 10, traj);
+
+  // Simplification results
+  auto simplifiedTraj = std::move(simplifiedPair.first);
+
+  // Check the first waypoint
+  auto s0 = stateSpace->createState();
+  simplifiedTraj->evaluate(0, s0);
+  auto r0 = s0.getSubStateHandle<R3>(0);
+  EXPECT_TRUE(r0.getValue().isApprox(startPose));
+
+  // Check the last waypoint
+  simplifiedTraj->evaluate(simplifiedTraj->getDuration(), s0);
+  r0 = s0.getSubStateHandle<R3>(0);
+  EXPECT_TRUE(r0.getValue().isApprox(goalPose));
+
+}
+
+
+// Test that the boolean returned is in agreement with shortening process
+TEST_F(SimplifierTest, ShortenThreeWayPointTraj)
+{
+
+  Eigen::Vector3d startPose(-5, -5, 0);
+  Eigen::Vector3d goalPose(5, 5, 0);
+
+  auto traj = planTrajForTest(stateSpace, interpolator, dmetric, sampler, collConstraint, 
+      boundsConstraint, boundsProjection, startPose, goalPose);
+
+  // Setup for simplification
+  aikido::distance::DistanceMetricPtr dmetric = aikido::distance::createDistanceMetric(stateSpace);
+  aikido::constraint::SampleablePtr sampler = aikido::constraint::createSampleableBounds(stateSpace, make_rng()); 
+  aikido::constraint::ProjectablePtr boundsProjection = aikido::constraint::createProjectableBounds(stateSpace);
+  aikido::constraint::TestablePtr boundsConstraint = aikido::constraint::createTestableBounds(stateSpace);
+  aikido::constraint::TestablePtr collConstraint = std::make_shared<MockTranslationalRobotConstraint>(
+        stateSpace, Eigen::Vector3d(-0.1, -0.1, -0.1),
+        Eigen::Vector3d(0.1, 0.1, 0.1));
+
+  auto simplifiedPair = aikido::planner::ompl::simplifyOMPL(
+      stateSpace, interpolator, std::move(dmetric), std::move(sampler), 
+      std::move(collConstraint), std::move(boundsConstraint), std::move(boundsProjection),
+      0.1, 5.0, 10, traj);
+
+  // Simplification results
+  auto simplifiedTraj = std::move(simplifiedPair.first);
+  bool shorten_success = simplifiedPair.second;
+
+  aikido::distance::DistanceMetricPtr Ddmetric = aikido::distance::createDistanceMetric(stateSpace);
+  double trajDistance = computeTrajLength<aikido::trajectory::InterpolatedPtr>(traj, stateSpace, Ddmetric);
+  double simplifiedTrajDistance = computeTrajLength<std::unique_ptr<aikido::trajectory::Interpolated>>(simplifiedTraj, stateSpace, Ddmetric);
+
+  EXPECT_TRUE(shorten_success == (simplifiedTrajDistance < trajDistance));  
```

Comment:
You raise an excellent point that shortcutting still could fail if we sample two points on the same segment.

Even so, I  would prefer to increase the number of iterations and `EXPECT_TRUE(shorten_success)`. One way of implementing that is to call `simplifyOMPL` in a loop up to `N` times with a generous `_timeLimit` and `_numAttempts = 1`. Adjust `N` until the tests passes with _very high_ probability.

Note that it doesn't take many iterations to effectively guarantee that `shorten_success` will be `true`. If the middle waypoint is halfway along the trajectory, every shortcutting attempt has a 50% probability of reducing the length of the trajectory. The probability that `N` sequential iterations will fail is `(0.5) ^ N`.